Documentation/proposals/202305-scrapeclasses.md

```diff
@@ -0,0 +1,223 @@
+kind: Probe
+spec:
+  scrapeClass: istio-mtls
+```
+
+### ServiceMonitor Resource
+
+Allow the user to select a scrape class for each endpoint.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+spec:
+  scrapeClass: istio-mtls
+  endpoints:
+  - port: http
+    path: /metrics
+```
+
+Out-of-scope: deprecation of the unsafe TLS settings in `ServiceMonitor`.
+
+### ScrapeConfig
+
+Allow the user to select a scrape class for the generic scrape configuration.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: ScrapeConfig
+metadata:
+  name: scrape-config
+spec:
+  scrapeClass: istio-mtls
+  staticConfigs:
+    [...]
+  httpSDConfig:
+    [...]
+  fileSDConfig:
+    [...]
+```
+
+## Test Plan
+
+1. Regression test; ensure no change to generated configs unless a scrape class is applied.
+2. Optionality of the default scrape class; ensure that it is an optional configuration element.
+3. Acceptance testing in Istio environment; ensure that the solution is effective for a key use case.
+
+## Alternatives
+
+### Global scrape TLS configuration
+
+An alternative solution would be to apply a default TLS configuration to all monitors.
+
+For example, via a hypothetical field `spec.scrapeTlsConfig`:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+spec:
+  scrapeTlsConfig:
+    caFile: "/etc/istio-certs/root-cert.pem"
+    certFile: "/etc/istio-certs/cert-chain.pem"
+    keyFile: "/etc/istio-certs/key.pem"
+    insecureSkipVerify: true
+```
+
+Objections:
+1. A singular default configuration may be too inflexible to effectively scrape a diverse set of pods. For example, in Istio,
+   some pods may be in STRICT mode.
+
+### Istio Permissive Mode
+
+An alternative for the Istio use case is to use PERMISSIVE mode (see [documentation](https://istio.io/latest/docs/concepts/security/#permissive-mode)),
+or to use `exclude` annotations on the pod such that the metrics endpoint bypasses mTLS.
+
+Objections:
+1. This solution is less secure at a transport level unless TLS is implemented at the application layer.
+2. Strict mTLS is the basis for [Istio authorization policy](https://istio.io/latest/docs/reference/config/security/authorization-policy/), e.g. to explicitly allow access from prom to the metrics endpoint.
+
+### ScrapeClass Resource
+
+A variant of the proposed solution is to introduce a new custom resource for defining scrape classes.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: ScrapeClass
+metadata:
+  name: istio-mtls
+spec:
+  tlsConfig:
+    caFile: "/etc/istio-certs/root-cert.pem"
+    certFile: "/etc/istio-certs/cert-chain.pem"
+    keyFile: "/etc/istio-certs/key.pem"
+    insecureSkipVerify: true
+```
+
+An open question is whether the resource would be cluster-scoped or namespace-scoped.
+
+Objections:
+1. Since the file paths are dependent on the volume mounts in the server, this approach may not achieve a meaningful decoupling.
+2. Extra complexity in defining a new CRD.
```

Comment:
I don't see a problem with volumeMounts here...

The operator could inject volume and volume mounts into the statefulset if needed, I don't see why using a CRD ould force the user to manually set volumes

Reply:
I think it could be quite complex.
Because the `Scrape Class` will not have any definition of the source of information.
Given the `tlsConfig` it's just pointing to which path the certificate is defined, I can't see Prometheus being able to mount volumes only based on that information and I'm not sure if adding a reference to volumes on the `Scrape Class` makes sense.

That said!
I can agree we have two touch points here and not 100% of decoupling, but it is what it is

Reply:
If we're defining TLS configuration in scrapeClasses, is there are reason to not specify the secret references here as well?

I couldn't understand why adding volume references in scrape classes wouldn't make sense 🤔